micropython/lora/lora/sx1276.py

```diff
@@ -0,0 +1,592 @@
+            self._buf2[1] = value
+            self._spi.write(self._buf2)
+            if _DEBUG:
+                dbg = hex(value)
+        else:  # value is a buffer
+            self._buf1[0] = reg | _WRITE_REG_BIT
+            self._spi.write(self._buf1)
+            self._spi.write(value)
+            if _DEBUG:
+                dbg = binascii.hexlify(value)
+        self._cs(1)
+
+        if _DEBUG:
+            print("W {:#x} ==> {}".format(reg, dbg))
+            self._reg_read(reg)  # log the readback as well
+
+    def _reg_update(self, reg, update_mask, new_value):
+        # Note: this function has a race condition, if the register value changes between
+        # read and write then old_value will be stale
+        old_value = self._reg_read(reg)
+        if update_mask:
+            value = ((old_value & ~update_mask) & 0xFF) | (new_value & update_mask)
+            self._reg_write(reg, value)
+        return old_value
+
+    def _reg_read(self, reg):
+        self._buf2[0] = reg
+        self._buf2[1] = 0xFF
+        self._cs(0)
+        self._spi.write_readinto(self._buf2, self._buf2)
+        self._cs(1)
+        if _DEBUG:
+            print("R {:#x} <== {:#x}".format(reg, self._buf2[1]))
+        return self._buf2[1]
+
+    def _reg_readinto(self, reg, buf):
+        self._cs(0)
+        self._spi.readinto(self._buf1, reg)
+        self._spi.readinto(buf)
+        if _DEBUG:
+            print("R {:#x} <== {}".format(reg, binascii.hexlify(buf)))
+        self._cs(1)
+
+    def _get_mode(self):
+        return self._reg_read(_REG_OPMODE) & _OPMODE_MODE_MASK
+
+    def _set_mode(self, mode, lora_en=True):
+        lora_val = _OPMODE_LONGRANGEMODE_LORA if lora_en else 0
+        old_value = self._reg_update(
+            _REG_OPMODE, _OPMODE_LONGRANGEMODE_LORA | _OPMODE_MODE_MASK, mode | lora_val
+        )
+        old_value &= _OPMODE_MODE_MASK
+        if old_value == _OPMODE_MODE_SLEEP:
+            time.sleep_ms(1)  # waking from sleep takes TS_OSC = 250us (typical)
+
+        if _DEBUG:
+            print("Mode {} -> {} ({:#x})".format(old_value, mode, self._reg_read(_REG_OPMODE)))
+
+        return old_value
+
+    def _dio_isr(self, pin):
+        if pin not in self._irq_en:
+            # TODO: This check against _irq_en shouldn't really be necessary,
+            # but weirdly DIO1 (configured for RxTimeout) seems to go high
+            # during TX, for no apparent reason (it's also masked out in
+            # _IRQ_FLAGS_MASK). Pull-up/down on the pin doesn't change it. This
+            # is probably a driver bug, and once it's fixed then _irq_en can be
+            # fully removed as the modem IRQ mask should determine which IRQs
+            # are driven high
+            if _DEBUG:
+                print("ISR ignored {}".format(pin))
+            return
+        if _DEBUG:
+            print("ISR {}".format(pin))
+        if self._irq_callback:
+            self._irq_callback()
```

Comment:
How is the callback supposed to distinguish successful TX, successful RX, RX with CRC error, and RX timeout? Why not pass it that info, which the driver already has?

Reply:
The purpose here is to wake up the code blocked in `receive()` or `transmit(), which then reads the interrupt status (and it already knows this information, as you note.)

Reply:
The IRQ callback still doesn't get any modem state, but I still think this is OK. What the implementer is supposed to do is now different and I hope clearer, and also documented.